go/ql/src/InconsistentCode/UnhandledCloseWritableHandle.ql

```diff
@@ -0,0 +1,160 @@
+ */
+predicate unhandledCall(DataFlow::CallNode call) {
+  exists(DeferStmt defer | defer.getCall() = call.asExpr()) or
+  exists(ExprStmt stmt | stmt.getExpr() = call.asExpr())
+}
+
+/**
+ * Determines whether `source` is a writable file handle returned by a `call` to the
+ * `os.OpenFile` function.
+ */
+predicate isWritableFileHandle(DataFlow::Node source, DataFlow::CallNode call) {
+  exists(OpenFileFun f, DataFlow::Node flags, QualifiedName flag |
+    // check that the source is a result of the call
+    source = call.getAResult() and
+    // find a call to the os.OpenFile function
+    f.getACall() = call and
+    // get the flags expression used for opening the file
+    call.getArgument(1) = flags and
+    // extract individual flags from the argument
+    // flag = flag.getAChild*() and
+    flag = getConstants(flags.asExpr()) and
+    // check for one which signals that the handle will be writable
+    // note that we are underestimating here, since the flags may be
+    // specified elsewhere
+    isWritable(flag.getTarget())
+  )
+}
+
+/**
+ * Determines whether `os.File.Close` is called on `sink`.
+ */
+predicate isCloseSink(DataFlow::Node sink, DataFlow::CallNode call) {
+  exists(CloseFileFun f |
+    // find calls to the os.File.Close function
+    f.getACall() = call and
+    // that are deferred
+    unhandledCall(call) and
+    // where the function is called on the sink
+    call.getReceiver() = sink
+  )
+}
+
+/**
+ * Determines whether `os.File.Sync` is called on `sink` and the result of the call is neither
+ * deferred nor discarded.
+ */
+predicate isHandledSync(DataFlow::Node sink, DataFlow::CallNode syncCall) {
+  // find a call of the `os.File.Sync` function
+  syncCall = any(SyncFileFun f).getACall() and
+  // match the sink with the object on which the method is called
+  syncCall.getReceiver() = sink and
+  // check that the result is neither deferred nor discarded
+  not unhandledCall(syncCall)
+}
+
+/**
+ * A data flow configuration which traces writable file handles resulting from calls to
+ * `os.OpenFile` to `os.File.Close` calls on them.
+ */
+class UnhandledFileCloseDataFlowConfiguration extends DataFlow::Configuration {
+  UnhandledFileCloseDataFlowConfiguration() { this = "UnhandledCloseWritableHandle" }
+
+  override predicate isSource(DataFlow::Node source) { isWritableFileHandle(source, _) }
+
+  override predicate isSink(DataFlow::Node sink) { isCloseSink(sink, _) }
+}
+
+/**
+ * Determines whether a `DataFlow::CallNode` is preceded by a call to `os.File.Sync`.
+ */
+predicate precededBySync(DataFlow::Node close, DataFlow::CallNode closeCall) {
+  // using the control flow graph, try to find a call to a handled call to `os.File.Sync`
+  // which precedes `closeCall`.
+  exists(IR::Instruction instr, DataFlow::Node syncReceiver, DataFlow::CallNode syncCall |
+    // find a predecessor to `closeCall` in the control flow graph
+    instr = closeCall.asInstruction().getAPredecessor*() and
```

Comment:
Is the control flow graph smart enough to put defer statements in the right place? I didn't realise that.

Reply:
I was surprised by that too. I wrote one of the test cases specifically because I was worried it wouldn't work right, only to find that it does.